File: tuan3/contro.py
def two_pointers_target_sum(a, target):
    # Buoc 1: Sap xep mang neu chua sap xep - O(n log n)
    arr = sorted(a)
    
    # Buoc 2: Kiem tra bang hai con tro - O(n)
    left = 0
    right = len(arr) - 1
    
    while left < right:
        current_sum = arr[left] + arr[right]
        if current_sum == target:
            return True, (arr[left], arr[right])
        elif current_sum < target:
            left += 1
        else:
            right -= 1
            
    return False, None

if __name__ == "__main__":
    A = [8, 3, 5, 1, 9, 2]
    S = 10  # Tim cap co tong bang 10 (vi du: 1 + 9 hoac 8 + 2)
    
    print(f"Mang dau vao: {A}")
    print(f"Tong can tim S = {S}")
    found, pair = two_pointers_target_sum(A, S)
    if found:
        print(f"Ket qua: Tim thay cap {pair} co tong bang {S}")
    else:
        print("Ket qua: Khong tim thay cap nao thoa man")

File: tuan3/contro.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Ham kiem tra cap phan tu co tong bang S su dung hai con tro
bool two_pointers_target_sum(vector<int> a, int target, pair<int, int> &result) {
    // Buoc 1: Sap xep mang - O(n log n)
    sort(a.begin(), a.end());

    // Buoc 2: Duyet hai con tro - O(n)
    int left = 0;
    int right = a.size() - 1;

    while (left < right) {
        int current_sum = a[left] + a[right];
        if (current_sum == target) {
            result = {a[left], a[right]};
            return true;
        } else if (current_sum < target) {
            left++;
        } else {
            right--;
        }
    }
    return false;
}

int main() {
    vector<int> A = {8, 3, 5, 1, 9, 2};
    int S = 10; // Vi du tim cap co tong bang 10 (1 + 9 hoac 2 + 8)

    cout << "Mang dau vao: ";
    for (int x : A) cout << x << " ";
    cout << "\nTong can tim S = " << S << "\n";

    pair<int, int> result;
    if (two_pointers_target_sum(A, S, result)) {
        cout << "Ket qua: Tim thay cap (" << result.first << ", " << result.second << ") co tong bang " << S << "\n";
    } else {
        cout << "Ket qua: Khong tim thay cap nao thoa man\n";
    }

    return 0;
}